bitmask-vpn: init at 0.21.6

ju1m opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Corresponding issue upstream: https://0xacab.org/leap/bitmask-vpn/-/issues/545

###### Motivation for this change
Being able to use `openvpn` through the graphical apps `riseup-vpn` (see https://riseup.net ) and/or `calyx-vpn` (see https://calyx.net ), both being instances of the https://bitmask.net tool by LEAP (LEAP Encryption Access Project).

Note that those rely upon the `bitmask-root` utility to setup the VPN, which currently only works with the `ip_tables` and `ip6_tables` kernel modules, not with `nf_tables`.

Note also that testing should be done by installing those packages via `environment.systemPackages = [ pkgs.riseup-vpn pkgs.calyx-vpn ];` in order to let NixOS install `bitmask-root`'s `polkit` action into `/run/current-system/sw/share/polkit-1/actions/`.

#### Testing this PR
Because those packages need to register a `polkit` action, they are not readily testable from `nixpkgs-review pr 137242` as one would expect. Instead, on NixOS, you can put this stanza in your `configuration.nix`:

```
environment.systemPackages = [
  (pkgs.libsForQt5.callPackage
    (pkgs.fetchurl {
      url = "https://raw.githubusercontent.com/ju1m/nixpkgs/bitmask/pkgs/tools/networking/bitmask-vpn/default.nix";
      hash = "sha256-mkHwuf1roF8BMORHR8bbe7wfFwPyk0NJgC/Q6KlE5jU";
      # This hash may change in case I update this PR, I'll update it in this intro message;
      # otherwise uncomment the fakeHash line and comment the line above
      # to force the fetching of the url again and get the latest hash of its content.
      #hash = lib.fakeHash;
    }) {
      provider = "riseup" /* or "calyx" */;
      inherit (pkgs.darwin.apple_sdk.frameworks) CoreFoundation Security;
    })
];
```
Then, after a `nixos-rebuild switch`, run `riseup-vpn` (resp. `calyx-vpn`).

###### Things done
- [X] Package https://0xacab.org/leap/bitmask-vpn
- [X] Test by running `riseup-vpn` from a console and using the systray menu.
- [ ] Darwin support.
```
$ riseup-vpn 
QSystemTrayIcon::setVisible: No Icon set
qrc:/qml/main.qml:41:9: QML QQuickItem: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
qrc:/qml/main.qml:108:9: QML QQuickItem: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
qrc:/qml/main.qml:177:9: QML BridgesItem: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
qml: Platform: linux
qml: DEBUG: Pre-seeded providers:
qml: {
    "default": "riseup",
    "providers": [
        {
            "apiURL": "https://api.black.riseup.net/",
            "applicationName": "RiseupVPN",
            "askForDonations": "true",
            "auth": "anon",
            "authEmptyPass": "false",
            "binaryName": "riseup-vpn",
            "caCertString": "-----BEGIN CERTIFICATE-----\nMIIFjTCCA3WgAwIBAgIBATANBgkqhkiG9w0BAQ0FADBZMRgwFgYDVQQKDA9SaXNl\ndXAgTmV0d29ya3MxGzAZBgNVBAsMEmh0dHBzOi8vcmlzZXVwLm5ldDEgMB4GA1UE\nAwwXUmlzZXVwIE5ldHdvcmtzIFJvb3QgQ0EwHhcNMTQwNDI4MDAwMDAwWhcNMjQw\nNDI4MDAwMDAwWjBZMRgwFgYDVQQKDA9SaXNldXAgTmV0d29ya3MxGzAZBgNVBAsM\nEmh0dHBzOi8vcmlzZXVwLm5ldDEgMB4GA1UEAwwXUmlzZXVwIE5ldHdvcmtzIFJv\nb3QgQ0EwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQC76J4ciMJ8Sg0m\nTP7DF2DT9zNe0Csk4myoMFC57rfJeqsAlJCv1XMzBmXrw8wq/9z7XHv6n/0sWU7a\n7cF2hLR33ktjwODlx7vorU39/lXLndo492ZBhXQtG1INMShyv+nlmzO6GT7ESfNE\nLliFitEzwIegpMqxCIHXFuobGSCWF4N0qLHkq/SYUMoOJ96O3hmPSl1kFDRMtWXY\niw1SEKjUvpyDJpVs3NGxeLCaA7bAWhDY5s5Yb2fA1o8ICAqhowurowJpW7n5ZuLK\n5VNTlNy6nZpkjt1QycYvNycffyPOFm/Q/RKDlvnorJIrihPkyniV3YY5cGgP+Qkx\nHUOT0uLA6LHtzfiyaOqkXwc4b0ZcQD5Vbf6Prd20Ppt6ei0zazkUPwxld3hgyw58\nm/4UIjG3PInWTNf293GngK2Bnz8Qx9e/6TueMSAn/3JBLem56E0WtmbLVjvko+LF\nPM5xA+m0BmuSJtrD1MUCXMhqYTtiOvgLBlUm5zkNxALzG+cXB28k6XikXt6MRG7q\nhzIPG38zwkooM55yy5i1YfcIi5NjMH6A+t4IJxxwb67MSb6UFOwg5kFokdONZcwj\nshczHdG9gLKSBIvrKa03Nd3W2dF9hMbRu//STcQxOailDBQCnXXfAATj9pYzdY4k\nha8VCAREGAKTDAex9oXf1yRuktES4QIDAQABo2AwXjAdBgNVHQ4EFgQUC4tdmLVu\nf9hwfK4AGliaet5KkcgwDgYDVR0PAQH/BAQDAgIEMAwGA1UdEwQFMAMBAf8wHwYD\nVR0jBBgwFoAUC4tdmLVuf9hwfK4AGliaet5KkcgwDQYJKoZIhvcNAQENBQADggIB\nAGzL+GRnYu99zFoy0bXJKOGCF5XUXP/3gIXPRDqQf5g7Cu/jYMID9dB3No4Zmf7v\nqHjiSXiS8jx1j/6/Luk6PpFbT7QYm4QLs1f4BlfZOti2KE8r7KRDPIecUsUXW6P/\n3GJAVYH/+7OjA39za9AieM7+H5BELGccGrM5wfl7JeEz8in+V2ZWDzHQO4hMkiTQ\n4ZckuaL201F68YpiItBNnJ9N5nHr1MRiGyApHmLXY/wvlrOpclh95qn+lG6/2jk7\n3AmihLOKYMlPwPakJg4PYczm3icFLgTpjV5sq2md9bRyAg3oPGfAuWHmKj2Ikqch\nTd5CHKGxEEWbGUWEMP0s1A/JHWiCbDigc4Cfxhy56CWG4q0tYtnc2GMw8OAUO6Wf\nXu5pYKNkzKSEtT/MrNJt44tTZWbKV/Pi/N2Fx36my7TgTUj7g3xcE9eF4JV2H/sg\ntsK3pwE0FEqGnT4qMFbixQmc8bGyuakr23wjMvfO7eZUxBuWYR2SkcP26sozF9PF\ntGhbZHQVGZUTVPyvwahMUEhbPGVerOW0IYpxkm0x/eaWdTc4vPpf/rIlgbAjarnJ\nUN9SaWRlWKSdP4haujnzCoJbM7dU9bjvlGZNyXEekgeT0W2qFeGGp+yyUWw8tNsp\n0BuC1b7uW/bBn/xKm319wXVDvBgZgcktMolak39V7DVO\n-----END CERTIFICATE-----",
            "donateURL": "https://riseup.net/vpn/donate",
            "geolocationAPI": "https://api.black.riseup.net:9001/json",
            "helpURL": "https://riseup.net/support",
            "name": "Riseup",
            "providerURL": "riseup.net",
            "timeStamp": "2021-09-10 01:55:47",
            "tosURL": "https://riseup.net/tos"
        }
    ]
}

qml: systray init completed
qml: show systray
2021/09/10 04:30:29 Client expects anon auth
2021/09/10 04:30:29 firewall stop
2021/09/10 04:30:30 Fetching gateways...
2021/09/10 04:30:31 Got sorted gateways: [zarapito.riseup.net hirondelle.riseup.net mouette.riseup.net hoatzin.riseup.net pie.riseup.net fournier.riseup.net hornero.riseup.net mockingjay.riseup.net shag.riseup.net redshank.riseup.net gaei.riseup.net yal.riseup.net starling.riseup.net limpkin.riseup.net crane.riseup.net garza.riseup.net]
2021/09/10 04:30:41 Fetching certificate to /tmp/leap-759395263/openvpn.pem
2021/09/10 04:30:42 DEBUG We have a valid cert: /tmp/leap-759395263/openvpn.pem
2021/09/10 04:30:42 Got sorted gateways: [zarapito.riseup.net fournier.riseup.net pie.riseup.net mouette.riseup.net hoatzin.riseup.net hirondelle.riseup.net mockingjay.riseup.net redshank.riseup.net shag.riseup.net hornero.riseup.net yal.riseup.net gaei.riseup.net starling.riseup.net crane.riseup.net limpkin.riseup.net garza.riseup.net]
2021/09/10 04:30:43 Got sorted gateways: [hoatzin.riseup.net pie.riseup.net zarapito.riseup.net mouette.riseup.net fournier.riseup.net hirondelle.riseup.net hornero.riseup.net redshank.riseup.net shag.riseup.net mockingjay.riseup.net gaei.riseup.net yal.riseup.net starling.riseup.net crane.riseup.net limpkin.riseup.net garza.riseup.net]
2021/09/10 04:30:43 firewall start
2021/09/10 04:30:43 openvpn start:  [--tun-ipv6 --auth SHA1 --cipher AES-128-CBC --keepalive 10 30 --tls-cipher DHE-RSA-AES128-SHA --remote 212.83.143.67 443 tcp4 --remote 212.83.146.228 443 tcp4 --remote 212.129.62.247 443 tcp4 --verb 3 --management-client --management 127.0.0.1 6061 --ca /tmp/leap-759395263/cacert.pem --cert /tmp/leap-759395263/openvpn.pem --key /tmp/leap-759395263/openvpn.pem --persist-tun]
2021/09/10 04:30:43 New connection into the management
2021/09/10 04:30:43 Event: INFO: OpenVPN Management Interface Version 3 -- type 'help' for more info
2021/09/10 04:30:43 Event: AUTH
2021/09/10 04:30:45 Event: GET_CONFIG
2021/09/10 04:30:45 Event: ASSIGN_IP: 10.41.0.9
2021/09/10 04:30:45 Event: CONNECTED: 212.83.143.67
2021/09/10 04:30:45 Connected to gateway: hoatzin.riseup.net
```
- [X] Test by running `calyx-vpn` from a console and using the systray.
There seems to be a mismatch in the certificate with the geolocation service, but mounting the VPN works nonetheless.
```
$ calyx-vpn 
QSystemTrayIcon::setVisible: No Icon set
qrc:/qml/main.qml:41:9: QML QQuickItem: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
qrc:/qml/main.qml:108:9: QML QQuickItem: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
qrc:/qml/main.qml:177:9: QML BridgesItem: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
qml: Platform: linux
qml: DEBUG: Pre-seeded providers:
qml: {
    "default": "calyx",
    "providers": [
        {
            "apiURL": "https://api.calyx.net:4430/",
            "applicationName": "CalyxVPN",
            "askForDonations": "true",
            "auth": "anon",
            "authEmptyPass": "false",
            "binaryName": "calyx-vpn",
            "caCertString": "-----BEGIN CERTIFICATE-----\nMIIFYzCCA0ugAwIBAgIBATANBgkqhkiG9w0BAQ0FADBEMQ4wDAYDVQQKDAVjYWx5\neDEaMBgGA1UECwwRaHR0cHM6Ly9jYWx5eC5uZXQxFjAUBgNVBAMMDWNhbHl4IFJv\nb3QgQ0EwHhcNMTMwNzAyMDAwMDAwWhcNMjMwNzAyMDAwMDAwWjBEMQ4wDAYDVQQK\nDAVjYWx5eDEaMBgGA1UECwwRaHR0cHM6Ly9jYWx5eC5uZXQxFjAUBgNVBAMMDWNh\nbHl4IFJvb3QgQ0EwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQDupdnx\nBgat537XOqrZOulE/RvjoXB1S07sy9/MMtksXFoQuWJZRCSTp1Jaqg3H/e9o1nct\nLQO91+izfJe07TUyajFl7CfllYgMeyKTYcT85dFwNX4pcIHZr8UpmO0MpGBoR4W1\n8cPa3vxAG0CsyUmrASJVyhRouk4qazRosM5RwBxTdMzCK7L3SwqPQoxlY9YmRJlD\nXYZlK5VMJd0dj9XxhMeFs5n43R0bsDENryrExSbuxoNfnUoQg3wffKk+Z0gW7YgW\nivPsbObqOgXUuBEU0xr9xMNBpU33ffLIsccrHq1EKp8zGfCOcww6v7+zEadUkVLo\n6j/rRhYYgRw9lijZG1rMuV/mTGnUqbjHsdoz5mzkFFWeTSqo44lvhveUyCcwRNmi\n2sjS77l0fCTzfreufffFoOEcRVMRfsnJdu/xPeARoXILEx8nQ421mSn6spOZlDQr\nTt0T0BAWt+VNc+m0IGSW3SwS7r5MUyQ/M5GrbQBGi5W2SzPriKZ79YTOwPVmXKLZ\nvJoEuKRDkEPJLBAhcD5oSQljOm/Wp/hjmRH4HnI1y4XMshWlDsyRDB1Au5yrsfwN\nnoFVSskEcbXlZfNgml4lktLBqz+qwsw+voq6Ak7ROKbc0ii5s8+iNMbAtIK7GcFF\nkuKKIyRmmGlDim/SDhlNdWo7Ah4Akde7zfWufwIDAQABo2AwXjAdBgNVHQ4EFgQU\nAY8+K4ZupAQ+L9ttFJG3vaLBq5gwDgYDVR0PAQH/BAQDAgIEMAwGA1UdEwQFMAMB\nAf8wHwYDVR0jBBgwFoAUAY8+K4ZupAQ+L9ttFJG3vaLBq5gwDQYJKoZIhvcNAQEN\nBQADggIBAOpXi5o3g/2o2rPa53iG7Zgcy8RpePGgZk6xknGYWeLamEqSh+XWQZ2w\n2kQP54bf8HfPj3ugJBWsVtYAs/ltJwzeBfYDrwEJd1N8tw2IRuGlQOWiTAVVLBj4\nZs+dikSuMoA399f/7BlUIEpVLUiV/emTtbkjFnDeKEV9zql6ypR0BtR8Knf8ALvL\nYfMsWLvTe4rXeypzxIaE2pn8ttcXLYAX0ml2MofTi5xcDhMn1vznKIvs82xhncQx\nI1MJMWqPHNHgJUJpA+y1IFh5LPbpag9PKQ0yQ9sM+/dyGumF2jElsMw71flh/Txr\n2dEv8+FNV1pPK26XJZBK24rNWFs30eAFfH9EQCwVla174I4PDoWqsIR7vtQMObDt\nBq34R3TjjJJIt2sCSlYLooWwiK7Q+d/SgYqA+MSDmmwhzm86ToK6cwbCsvuw1AxR\nX6VIs4U8wOotgljzX/CSpKqlxcqZjhnAuelZ1+KiN8RHKPj7AzSLYOv/YwTjLTIq\nEOxquoNR58uDa5pBG22a7xWbSaKosn/mEl8SrUr6klzzc8Vh09IMoxrw74uLdAg2\n1jnrhm7qg91Ttb0aXiqbV+Kg/qQzojdewnnoBFnv4jaQ3y8zDCfMhsBtWlWz4Knb\nZqga1WyRm3Gj1j6IV0oOincYMrw5YA7bgXpwop/Lo/mmliMA14ps\n-----END CERTIFICATE-----",
            "donateURL": "",
            "geolocationAPI": "https://api.black.riseup.net:9001/json",
            "helpURL": "https://calyx.net/support",
            "name": "Calyx",
            "providerURL": "https://calyx.net",
            "timeStamp": "2021-09-10 02:18:57",
            "tosURL": "https://calyx.net/tos"
        }
    ]
}

qml: systray init completed
qml: show systray
2021/09/10 05:23:40 Client expects anon auth
2021/09/10 05:23:40 firewall stop
2021/09/10 05:23:40 Fetching gateways...
2021/09/10 05:23:42 ERROR: could not fetch geolocation: Post "https://api.black.riseup.net:9001/json": x509: certificate signed by unknown authority
2021/09/10 05:23:58 Fetching certificate to /tmp/leap-097269187/openvpn.pem
2021/09/10 05:23:59 DEBUG We have a valid cert: /tmp/leap-097269187/openvpn.pem
2021/09/10 05:24:00 ERROR: could not fetch geolocation: Post "https://api.black.riseup.net:9001/json": x509: certificate signed by unknown authority
2021/09/10 05:24:01 ERROR: could not fetch geolocation: Post "https://api.black.riseup.net:9001/json": x509: certificate signed by unknown authority
2021/09/10 05:24:01 firewall start
2021/09/10 05:24:02 openvpn start:  [--tls-cipher DHE-RSA-AES128-SHA --tun-ipv6 --auth SHA1 --cipher AES-128-CBC --keepalive 10 30 --remote 162.247.73.193 443 tcp4 --verb 3 --management-client --management 127.0.0.1 6061 --ca /tmp/leap-097269187/cacert.pem --cert /tmp/leap-097269187/openvpn.pem --key /tmp/leap-097269187/openvpn.pem --persist-tun]
2021/09/10 05:24:02 New connection into the management
2021/09/10 05:24:02 Event: INFO: OpenVPN Management Interface Version 3 -- type 'help' for more info
2021/09/10 05:24:02 Event: AUTH
2021/09/10 05:24:08 Event: GET_CONFIG
2021/09/10 05:24:08 Event: ASSIGN_IP: 10.41.2.64
2021/09/10 05:24:08 Event: CONNECTED: 162.247.73.193
2021/09/10 05:24:08 Connected to gateway: vpn2.calyx.net
```

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ju1m committed:
bitmask-vpn: init at 0.21.6

r-rmcgibbo commented:
Result of `nixpkgs-review pr 137247` at c234a58e run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>calyx-vpn</li>
    <li>riseup-vpn</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/585

SuperSandro2000 commented on pkgs/tools/networking/bitmask-vpn/default.nix:
```diff
@@ -0,0 +1,149 @@
+    which
+    python3Packages.wrapPython
+    qttools
+    # make installer is not used
+    #qtinstaller
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/networking/bitmask-vpn/default.nix:
```diff
@@ -0,0 +1,149 @@
+  doCheck = false;
+
+  meta = {
+    description = ''
```
```suggestion
    description = "...";
    longDescription = ''
```

ju1m commented:
@SuperSandro2000, thanks, I've applied your suggestions.

happysalada commented:
Result of `nixpkgs-review pr 137247` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>calyx-vpn</li>
    <li>riseup-vpn</li>
  </ul>
</details>

happysalada commented on pkgs/tools/networking/bitmask-vpn/default.nix:
```diff
@@ -0,0 +1,148 @@
+, provider ? "riseup"
+}:
+let
+  version = "0.21.6";
```
latest seems to be 0.21.8 https://0xacab.org/leap/bitmask-vpn/-/blob/main/CHANGELOG

ju1m replied:
@happysalada , indeed a changelog entry is there, and the milestone is past due, but there is no tag yet: https://0xacab.org/leap/bitmask-vpn/-/tags

happysalada commented:
Is this not supposed to work on darwin at all ?
I'm just wondering if there is a way to make it work.
I couldn't find any reference to osx on the repo in the docs, but there seemed to be a commit message that mentioned it.

ju1m commented:
> Is this not supposed to work on darwin at all ?
> I'm just wondering if there is a way to make it work.
> I couldn't find any reference to osx on the repo in the docs, but there seemed to be a commit message that mentioned it.

@happysalada, looks like there is some [Makefile rules for the darwin platform](https://0xacab.org/search?search=darwin&group_id=52&project_id=1042&scope=&search_code=true&snippets=false&repository_ref=&nav_source=navbar) but I do not have such platform to test them. If you're able to build in `darwin` I guess submitting your fix would work by opening a PR against https://github.com/ju1m/nixpkgs/tree/bitmask

happysalada commented on pkgs/tools/networking/bitmask-vpn/default.nix:
```diff
@@ -0,0 +1,148 @@
+    wrapPythonProgramsIn branding/scripts
+  '';
+
+  nativeBuildInputs = [
```
it seems that most of those inputs are only used for linux.
Could you mark them as optional for linux for now.
https://0xacab.org/leap/bitmask-vpn/-/blob/main/Makefile#L78

I should have time to test on darwin on monday.

ju1m replied:
@happysalada, I don't think it's reliable to compare Debian's dependencies and Brew's dependencies: a single Brew package may provide the same content provided by multiple Debian packages.
I've read the `Makefile` to see whether the building toolchain would be significantly different on `darwin` and those `darwin`-specific rules appear to be mostly about building an installer (dmg image), but I guess that with Nixpkgs we would not need such installer. But I don't know `darwin` so I may be wrong.
Best is to make it work on `darwin` if you can, and then see if some dependencies are superfluous or broken but optional there.
You may use a developing cycle based upon:

```
rm -rf source
nix -L develop -f . riseup-vpn
unpackPhase
cd source
patchPhase
eval "$buildPhase" || echo ERROR
[hack ... hack ... hack]
```
And that would be `nix -L build -f . riseup-vpn` to get the final `./result/`.
Do not hesitate to post the errors you cannot overcome here, so I can see if I can help.

ju1m commented:
- [X] Add a `$out/share/applications/${provider}-vpn.desktop`, that should help end-users to launch the VPN (not tested by me though).

tomberek reviewed (changes requested):
Functions as expected on NixOS. Did not test graphical or QT portions, only VPN connectivity and routing.

Recommend adding a few statements or comments about how this is expected to function and how users should install it.

happysalada commented:
I have had a chance to test on darwin, here are the problems.

- `which: no binarycreator` I am not sure where binarycreator is supposed to be found. I think it's just a matter of doing a substitution, but don't know where the package bin is supposed to be
- `iproute2` and `iptables` are packages specific to linux, so their substitution needs to be conditional on linux.
- missing debug libs
```
clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
/libQt5Widgets_debug.dylib'
clang-7: error: no such file or directory: '/nix/store/i3ff2saa275pvcrz57gr4q0dz371czm5-qtdeclarative-5.12
.10/lib/libQt5Quick_debug.dylib'
clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
/libQt5Gui_debug.dylib'
clang-7: error: no such file or directory: '/nix/store/i3ff2saa275pvcrz57gr4q0dz371czm5-qtdeclarative-5.12
.10/lib/libQt5Qml_debug.dylib'
clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
/libQt5Network_debug.dylib'
clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
/libQt5Core_debug.dylib'
```
I'm not sure how to include those debug libs, I've tried adding `qtbase.dev` and `qtdeclarative.dev` in buildInputs but it didn't change anything.
- There are some missing symbols
```
Undefined symbols for architecture x86_64:
  "_CFArrayGetCount", referenced from:
      _crypto/x509/internal/macos.x509_CFArrayGetCount_trampoline in libgoshim.a(go.o)
  "_CFArrayGetValueAtIndex", referenced from:
      _crypto/x509/internal/macos.x509_CFArrayGetValueAtIndex_trampoline in libgoshim.a(go.o)
  "_CFDataGetBytePtr", referenced from:
      _crypto/x509/internal/macos.x509_CFDataGetBytePtr_trampoline in libgoshim.a(go.o)
  "_CFDataGetLength", referenced from:
      _crypto/x509/internal/macos.x509_CFDataGetLength_trampoline in libgoshim.a(go.o)
  "_CFDictionaryGetValueIfPresent", referenced from:
      _crypto/x509/internal/macos.x509_CFDictionaryGetValueIfPresent_trampoline in libgoshim.a(go.o)
  "_CFEqual", referenced from:
      _crypto/x509/internal/macos.x509_CFEqual_trampoline in libgoshim.a(go.o)
  "_CFNumberGetValue", referenced from:
      _crypto/x509/internal/macos.x509_CFNumberGetValue_trampoline in libgoshim.a(go.o)
  "_CFRelease", referenced from:
      _crypto/x509/internal/macos.x509_CFRelease_trampoline in libgoshim.a(go.o)
  "_CFStringCreateWithBytes", referenced from:
      _crypto/x509/internal/macos.x509_CFStringCreateWithBytes_trampoline in libgoshim.a(go.o)
ld: symbol(s) not found for architecture x86_64
```
those are in the CoreFoundation framework, it seems though that a new flag needs to passed to clang. Is the right way to set the `NIX_LDFLAGS` variables ? I'm not sure. The flag I want to pass is `-framework CoreFoundation`.

I have time to test this everyday if needed. I realise though that you might want to split darwin support in a separate PR, if you would rather do that, I understand.

ju1m commented:
> Functions as expected on NixOS. Did not test graphical or QT portions, only VPN connectivity and routing.
> 
> Recommend adding a few statements or comments about how this is expected to function and how users should install it.

@tomberek, thanks for the review! Where would you like me to put such comments?
Now that there is a `/run/current-system/sw/share/applications/riseup-vpn.desktop`, `riseup-vpn` should appear amongst other applications in the desktop manager's XDG menu. Note that the application appears as a systray icon.

ju1m commented:
> I have had a chance to test on darwin, here are the problems.
> 
>     * `which: no binarycreator` I am not sure where binarycreator is supposed to be found. I think it's just a matter of doing a substitution, but don't know where the package bin is supposed to be

@happysalada, `binarycreator` is in `qtinstaller`, it raises a warning because the strict `:=` is used instead of the lazy `=` in the `Makefile`, but is only needed by the `installer` target which is not used at this point. As documented in `docs/build-osx.rst`, `installer` is used to build the `.dmg` but I _guess_ that's not Nixpkgs's way to install on `Darwin`.
> 
>     * `iproute2` and `iptables` are packages specific to linux, so their substitution needs to be conditional on linux.

I've guarded uses of `bitmask-root` behind a `stdenv.isLinux` (AFAIU, `pkg/helper` is used on `Windows` and `Darwin`). Thanks!

>     * missing debug libs
> ```
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Widgets_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/i3ff2saa275pvcrz57gr4q0dz371czm5-qtdeclarative-5.12
> .10/lib/libQt5Quick_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Gui_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/i3ff2saa275pvcrz57gr4q0dz371czm5-qtdeclarative-5.12
> .10/lib/libQt5Qml_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Network_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Core_debug.dylib'
> ```
> 
> I'm not sure how to include those debug libs, I've tried adding `qtbase.dev` and `qtdeclarative.dev` in buildInputs but it didn't change anything.

Hm, I've no idea, sometimes it works better in `nativeBuildInputs`.

>     * There are some missing symbols
> ```
> Undefined symbols for architecture x86_64:
>   "_CFArrayGetCount", referenced from:
>       _crypto/x509/internal/macos.x509_CFArrayGetCount_trampoline in libgoshim.a(go.o)
>   "_CFArrayGetValueAtIndex", referenced from:
>       _crypto/x509/internal/macos.x509_CFArrayGetValueAtIndex_trampoline in libgoshim.a(go.o)
>   "_CFDataGetBytePtr", referenced from:
>       _crypto/x509/internal/macos.x509_CFDataGetBytePtr_trampoline in libgoshim.a(go.o)
>   "_CFDataGetLength", referenced from:
>       _crypto/x509/internal/macos.x509_CFDataGetLength_trampoline in libgoshim.a(go.o)
>   "_CFDictionaryGetValueIfPresent", referenced from:
>       _crypto/x509/internal/macos.x509_CFDictionaryGetValueIfPresent_trampoline in libgoshim.a(go.o)
>   "_CFEqual", referenced from:
>       _crypto/x509/internal/macos.x509_CFEqual_trampoline in libgoshim.a(go.o)
>   "_CFNumberGetValue", referenced from:
>       _crypto/x509/internal/macos.x509_CFNumberGetValue_trampoline in libgoshim.a(go.o)
>   "_CFRelease", referenced from:
>       _crypto/x509/internal/macos.x509_CFRelease_trampoline in libgoshim.a(go.o)
>   "_CFStringCreateWithBytes", referenced from:
>       _crypto/x509/internal/macos.x509_CFStringCreateWithBytes_trampoline in libgoshim.a(go.o)
> ld: symbol(s) not found for architecture x86_64
> ```
> those are in the CoreFoundation framework, it seems though that a new flag needs to passed to clang. Is the right way to set the `NIX_LDFLAGS` variables ? I'm not sure. The flag I want to pass is `-framework CoreFoundation`.

I would try to patch `bitmask.pro` where there's already a `-framework` flag, but this seems very strange that such flag is missing.
But `git grep -i corefoundation` suggests rather to add `CoreFoundation` to `buildInputs`. i've done that and also added `Security` (which appears in `bitmask.pro`).

> I have time to test this everyday if needed. I realise though that you might want to split darwin support in a separate PR, if you would rather do that, I understand.

Since I am unable to develop on Darwin, it would indeed be better for me to pass the relay and let you shepherd `Darwin` support on a following PR. In the meantime I've added what I can to improve the state of this package on `Darwin` based on our discussions here.

tomberek commented:
> Now that there is a `/run/current-system/sw/share/applications/riseup-vpn.desktop`, `riseup-vpn` should appear amongst other applications in the desktop manager's XDG menu. Note that the application appears as a systray icon.

I don't run with a normal desktop manager, nor have a systray.... so this is hard for me to test in my native setup. As for the comments, they can go into meta or just raw Nix comments.

ju1m commented:
> I don't run with a normal desktop manager, nor have a systray.... so this is hard for me to test in my native setup.

@tomberek, hm, I tried to use `riseup-vpn --start-vpn on` but to no avail, log strangely says:
```
2021/09/16 16:11:56 -start-vpn should be 'on' or 'off'
```
So I don't see how you could use `riseup-vpn` without using the systray.
FYI, Riseup's VPN can be used without that client, for instance I've put the config I'm using in the intro message of https://github.com/NixOS/nixpkgs/pull/109643 

> As for the comments, they can go into meta or just raw Nix comments.

I've added comments to `meta.longDescription`.

happysalada commented:
@ju1m thanks for the tips, let me give this one more go at it tomorrow.

SuperSandro2000 commented on pkgs/tools/networking/bitmask-vpn/default.nix:
```diff
@@ -0,0 +1,158 @@
+    qtbase
+    qmltermwidget
+    qtdeclarative
+  ] ++ lib.optionals (stdenv.isDarwin) [ CoreFoundation Security ];
```
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [ CoreFoundation Security ];
```

SuperSandro2000 commented on pkgs/tools/networking/bitmask-vpn/default.nix:
```diff
@@ -0,0 +1,158 @@
+    # Remove timestamps in comments
+    sed -i -e '/^\/\//d' pkg/config/version/version.go
+
+    make build
```
```suggestion
    make build -j$NIX_BUILD_CORES
```

ju1m replied:
@SuperSandro2000, thanks for the suggestion, unfortunately the `Makefile`'s rules are not thread-safe; I've added a comment to explain why `-j` can't be used here.

happysalada commented:
I actually made some progress.
@ju1m thanks for the tip, adding the framework in the bitmask.pro fixed the framework error. I messaged upstream.

Regarding the other error with missing debug symbols for qt.
I tried building qtbase with debug symbols and I ran into this error https://bugreports.qt.io/browse/QTBUG-76777
which is basically that this cannot be built with the version of apple sdk we are using. So until sdk is upgraded this won't build for darwin. As far as I'm concerned then we can leave it like this.

Thanks again for your help.

ju1m commented:
@happysalada, maybe it could work by using `darwin.apple_sdk_11_0` instead of `darwin.apple_sdk` (passed in the `all-packages.nix`)? Or would it only work on `Aarch64`?

happysalada commented:
I tried before to get the apple_11_0 sdk on x86_64 without success.
only the "apple_sdk" attribute is available from `darwin` (can be verified in a repl), and that attribute resolve automatically to the old sdk for non aarch64 architecture. If I find a way to resolve this, I'll update here.

happysalada commented on pkgs/tools/networking/bitmask-vpn/default.nix:
```diff
@@ -0,0 +1,166 @@
+    '';
+    homepage = "https://bitmask.net";
+    license = lib.licenses.gpl3Plus;
+    maintainers = with lib.maintainers; [ julm ];
```
can we just add a `platforms = lib.platforms.linux;` for now.

happysalada replied:
Perhaps we can add a comment on top with the following
`# darwin requires macos sdk 10.13`

ju1m replied:
Done

happysalada commented:
Thank you!
